pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -916,33 +995,43 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 	stat.ProcessingTime.End(true)
 	stat.UpstreamTime.Start()
 
+	var ipFamily ipfamily.IPFamily
+	if targetServerAddr.Is4() {
+		ipFamily = ipfamily.IPv4()
+	} else {
+		ipFamily = ipfamily.IPv6()
+	}
+
 	dialer := net.Dialer{
 		Timeout: 2 * time.Second,
 		Control: func(network, address string, c syscall.RawConn) error {
 			var soerr error
 			if err := c.Control(func(su uintptr) {
-				soerr = setSoMark(int(su), ep.GetIdentity())
+				soerr = setSoMarks(int(su), ipFamily, ep.GetIdentity())
 			}); err != nil {
 				return err
 			}
 			return soerr
 		},
 	}
-	client.Dialer = &dialer
 
-	conn, err := client.Dial(targetServerAddrStr)
-	if err != nil {
-		err := fmt.Errorf("failed to dial connection to %v: %w", targetServerAddrStr, err)
-		stat.Err = err
-		scopedLog.WithError(err).Error("Failed to dial connection to the upstream DNS server, cannot service DNS request")
-		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerID, targetServerAddrStr, request, protocol, false, &stat)
-		p.sendRefused(scopedLog, w, request)
-		return
+	var key string
+	if !ep.IsHost() && !epAddr.IsLoopback() && !targetServerIsWorld && !targetServerID.HasLocalScope() && !targetServerIsLocal {
+		dialer.LocalAddr = originalSrcAddr
+		key = protocol + "-" + originalSrcAddr.String() + "-" + targetServerAddrStr
+	}
```

Comment:
I think it would be important to leave a comment on these lines. They are kind of the more relevant lines in the entire PR.

Reply:
Added back the comment, even though the comment was (and still is) inverted in its logic vs. the code below. But it may be helpful to think about this from both angles!